File: #132_Palindrome_Partition_II/Solution.java
public class Solution {
    public int minCut(String s) {
        if (s == null) return 0;
        int len = s.length();
        boolean[][] isPalindrome = new boolean[len + 1][len + 1];
        int[] dp = new int[len];
        /* initialize */
        for (int i = 0; i < len; i++) {
        	dp[i] = i;
        }
        for (int i = 0; i < len + 1; i++) {
        	isPalindrome[i][i] = true;
        }
        for (int j = 1; j < len + 1; j++) {
        	for (int i = 0; i < j; i++) {
    			isPalindrome[i][j] = (s.charAt(i) == s.charAt(j - 1) && ((j - i == 1) || isPalindrome[i + 1][j - 1]));
    			if (isPalindrome[i][j]) {
	    			if (i == 0) dp[j - 1] = 0;
    				else dp[j - 1] = (1 + dp[i - 1] < dp[j - 1]) ? 1 + dp[i - 1] : dp[j - 1];
    			}
        	}
        }
        return dp[len - 1];
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	System.out.println(sol.minCut("cdd"));
    }
}

File: #132_Palindrome_Partition_II/Solution.py
'''
O(n ^ 2) time Solution
'''

class Solution:
	# @param {string} s
	# @return {integer}
	def minCut(self, s):
		if not s:
			return 0
		dp = [[False for j in range(len(s))] for i in range(len(s))]
		for x in range(len(s)):
			dp[x][x] = True
		for j in range(len(s)):
			for i in range(len(s)):
				if i >= j:
					break
				elif j - i == 1 and s[i] == s[j]:
					dp[i][j] = True
				elif s[i] == s[j] and dp[i + 1][j - 1]:
					dp[i][j] = True
		cut = [len(s) for x in range(len(s) + 1)]
		cut[0] = -1
		for x in range(len(s)):
			for y in range(x + 1):
				if dp[y][x]:
					cut[x + 1] = min(cut[y] + 1, cut[x + 1])
		return cut[len(s)]

sol = Solution()
print sol.minCut("a")